### PR TITLE
[WIP][DRAFT] Pip installable PySpark

### PR DESCRIPTION
**Pull request by holdenk**
Draft pip installable PySpark package. This does a bunch of hacks to copy the jars over.
Done:
- pip installable on conda [manual tested]
- setup.py installed on a non-pip managed system (RHEL) with YARN [manual tested]
- Automated testing of this (virtualenv)
- packaging and signing with release-build*

Possible follow up work:
- release-build update to publish to PyPi (not being a commiter difficult to work on)
- investigate details of wheel caching
- consider how we want to number our dev/snapshot versions
- Windows? idk

Explicitly out of scope:
- Using pip installed PySpark to start a standalone cluster
- Using pip installed PySpark for non-Python Spark programs

*I've done some hacks to test release-build locally but as a non-committer I've just done local testing.

### CODE REVIEW

File: bin/find-spark-home [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+if [ ! -z "${SPARK_HOME}" ]; then
+   exit 0
+elif [ ! -f $FIND_SPARK_HOME_PYTHON_SCRIPT ]; then
+  # If we are not in the same directory as find_spark_home.py we are not pip installed
+  export SPARK_HOME="$(cd "`dirname "$0"`"/..; pwd)"
+else
+  # We are pip installed, use the Python script to resolve a reasonable SPARK_HOME
+  # Default to standard python interpreter unless told otherwise
+  if [[ -z "$PYSPARK_DRIVER_PYTHON" ]]; then
```
Comment (hougs) [inline bin/find-spark-home:37]: Doesn't this already happen in the python runner?

File: bin/pyspark [hunk 2]
```diff
@@ -61,7 +61,12 @@ export PYTHONPATH="${SPARK_HOME}/python/lib/py4j-0.10.4-src.zip:$PYTHONPATH"
 
 # Load the PySpark shell.py script when ./pyspark is used interactively:
 export OLD_PYTHONSTARTUP="$PYTHONSTARTUP"
-export PYTHONSTARTUP="${SPARK_HOME}/python/pyspark/shell.py"
+# Check if we are pip installed or otherwise
```
Comment (hougs) [inline bin/pyspark]: Why is this necessary? Could this be a single code path? That would be better if possible.

File: bin/pyspark2.cmd [hunk 3]
```diff
@@ -33,6 +33,11 @@ set PYTHONPATH=%SPARK_HOME%\python;%PYTHONPATH%
 set PYTHONPATH=%SPARK_HOME%\python\lib\py4j-0.10.4-src.zip;%PYTHONPATH%
 
 set OLD_PYTHONSTARTUP=%PYTHONSTARTUP%
+rem Check if we are pip installed or not
+IF EXIST %SPARK_HOME%\python\pyspark\shell.py (
```
Comment (hougs) [inline bin/pyspark2.cmd]: again with the double code path. This seems messy and hard to maintain.

**Review by hougs — CHANGES_REQUESTED** [review]
I had a quick glance at a few changes. The manual 'is this pip installed or not' checks are messy and hard to maintain. I think having the paths be consistent no matter the method of installation is far preferable. Can it be done? What is the curren challenge?

**Comment by holdenk** [comment]
So we could set the SPARK_HOME to be a level up, but then our package names wouldn't make sense (e.g. we'd have to publish a jar package which might conflict with someone elses package named jar). Alternatively we could have a copy of all the code, which might work (since the spark scripts will look under SPARK_HOME/python/pyspark), but it feels a bit weird to have a copy of everything. For the shell.py it maybe makes sense to just copy the shell script since it is the only script which is referenced by path (everything else is by import which will work just fine).

File: bin/find-spark-home (hunk 1 above)
Comment (holdenk) [reply bin/find-spark-home:37]: It does - so the reasoning behind this is that this is called before the Python runner and if the user has a specific driver Python they want to use we should resolve PySpark's location using that (e.g. if the user has both Python 2 and Python 3 envs setup). We could probably 86 this and trust the user to have the preffered Python first in path (e.g. by activating the correct virtualenv) but I figured it was safer to respect PYSPARK_DRIVER_PYTHON if set than not.

**Comment by holdenk** [comment]
Hi @minrk I figured I'd ping you (as suggested by @rgbkrk ) to see if you maybe have any thoughts or suggestions around packaging PySpark this way. This isn't yet the "formal" pull request but unless I've done something really off the bend I'll probably make it tomorrow morning before the second day of Spark Summit EU (and can continue discussion on the "formal" PR of course too).

Packaging the jars is a little awkward as is trying to provide an env that both makes sense with the existing shell scripts and python based import so if there are ways to do it better I'd love to know (but I also have the impression that there might not be a lot of better ways to do this).

**Comment by minrk** [comment]
@holdenk thanks, this is awesome! So the jars are copied into the Python package and bundled together on install, meaning a pip-installed pyspark doesn't refer to any spark files outside itself?

I'm sure there are a dozen ways to do it, depending on how/when you want to tie pyspark to specific jars.

File: bin/find-spark-home [hunk 4]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
```
Comment (minrk) [inline bin/find-spark-home:1]: Big fan of find-spark-home as a standalone entrypoint.

File: docs/index.md [hunk 5]
```diff
@@ -14,7 +14,9 @@ It also supports a rich set of higher-level tools including [Spark SQL](sql-prog
 
 Get Spark from the [downloads page](http://spark.apache.org/downloads.html) of the project website. This documentation is for Spark version {{site.SPARK_VERSION}}. Spark uses Hadoop's client libraries for HDFS and YARN. Downloads are pre-packaged for a handful of popular Hadoop versions.
 Users can also download a "Hadoop free" binary and run Spark with any Hadoop version
-[by augmenting Spark's classpath](hadoop-provided.html). 
+[by augmenting Spark's classpath](hadoop-provided.html).
+Scala and Java users can include Spark in their projects using it's maven cooridnates and Python users can also install Spark from PyPi.
```
Comment (minrk) [inline docs/index.md]: typo: PyPI (Python Package Index), not PyPi

File: python/setup.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        os.symlink(JARS_PATH, JARS_TARGET)
+        os.symlink(SCRIPTS_PATH, SCRIPTS_TARGET)
+        os.symlink(EXAMPLES_PATH, EXAMPLES_TARGET)
+    else:
+        # We add find_spark_home.py to the bin directory we install so that pip installed PySpark
+        # will search for SPARK_HOME with Python.
+        # We only do this copy when we aren't inside of Spark (e.g. the packaging tool has copied
+        # all the files into a temp directory) since otherwise the copy would go into the symlinked
+        # directory.
+        copyfile("pyspark/find_spark_home.py", SCRIPTS_TARGET + "/find_spark_home.py")
```
Comment (minrk) [inline python/setup.py]: You could avoid the copied file by having a script that imports:

``` python
#!/usr/bin/env python
if __name__ == '__main__':
    from pyspark.find_spark_home import main
    main()
```

and you should be able to avoid copying into the symlinks directory by having two sources of scripts: the symlinked SCRIPTS_TARGET and a local directory containing scripts like this one.

File: python/setup.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    except:
+        print("Temp path for symlink to parent already exists %s" % TEMP_PATH, file=sys.stderr)
+        exit(-1)
+
+try:
+    if (in_spark):
+        # Construct the symlink farm
+        os.symlink(JARS_PATH, JARS_TARGET)
+        os.symlink(SCRIPTS_PATH, SCRIPTS_TARGET)
+        os.symlink(EXAMPLES_PATH, EXAMPLES_TARGET)
```
Comment (minrk) [inline python/setup.py]: Depending on your Windows plans, you may need to handle the fact that symlink is often not available on Windows.

Comment (holdenk) [reply python/setup.py]: I think it's probably ok to only support packaging on Unix systems if I directly call it out - but thats a good catch (could also do a fall back to copy since there isn't really a lot io being saved anyways).

**Comment by holdenk** [comment]
@minrk thanks for taking a look and providing these suggestions :) So yah currently the jars are copied from the scala build and the Python package doesn't need to depend on any other Spark install outside of it. I figured this was also safer since if we had it pip installed and JARs provided elsewhere I could see people accidentally mixing and matching the two versions and Py4J error messages aren't the simplest to understand at the best of times. I'll work on this some more tomorrow after the keynotes :)

File: python/setup.py [hunk 8]
```diff
@@ -23,6 +23,11 @@
 from setuptools import setup, find_packages
 from shutil import copyfile, copytree, rmtree
 
+if sys.version_info < (2, 7):
+    print("Python versions prior to 2.7 are not supported for pip installed PySpark.",
+          file=sys.stderr)
+    exit(-1)
```
Comment (rgbkrk) [inline python/setup.py:29]: 👍
